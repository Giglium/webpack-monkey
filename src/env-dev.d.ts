/// <reference types="webpack/module" />
/// <reference types="webpack-dev-server" />

declare const __PORT__: number
