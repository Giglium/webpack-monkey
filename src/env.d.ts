declare namespace webpack {
  interface Hot {
    monkeyReload(options?: import("./client/hmr").MonkeyReloadOptions): void
  }
}
