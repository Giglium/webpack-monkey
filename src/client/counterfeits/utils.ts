export function parseHeaders(headers: string): Record<string, string> {
  return headers
    .split("\n")
    .filter(Boolean)
    .map((line) => line.split(": "))
    .reduce((headers, [name, value]) => {
      headers[name!] = value!
      return headers
    }, {} as Record<string, string>)
}
