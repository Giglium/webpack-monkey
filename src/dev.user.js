;(async function () {
  "use strict"

  const { clientScript, runtimeScript } = __MK_DEV_INJECTION__

  if (!clientScript || !runtimeScript) {
    console.warn("[Monkey Dev] client script not found")
    return
  }

  window.__MK_GLOBAL__ = unsafeWindow.__MK_GLOBAL__ = {
    loadScript: (url) => {
      return new Promise((resolve, reject) => {
        const fail = () => {
          reject(
            new Error(`failed to load script (${url}), please check if the server is running.`)
          )
        }

        GM_xmlhttpRequest({
          method: "GET",
          url,
          onload: (res) => {
            try {
              if (res.status === 404) {
                throw new Error(`Script not found (${url}).`)
              }

              const fakeScript = document.createElement("script")
              fakeScript.src = url

              Object.defineProperty(document, "currentScript", {
                get: () => fakeScript,
                configurable: true,
              })

              eval(res.responseText)

              delete document.currentScript

              resolve()
            } catch (e) {
              reject(e)
            }
          },
          onerror: fail,
          ontimeout: fail,
        })
      })
    },
  }

  console.log("[Monkey Dev] running")

  __MK_GLOBAL__.loadScript(runtimeScript).then(() => __MK_GLOBAL__.loadScript(clientScript))
})()
