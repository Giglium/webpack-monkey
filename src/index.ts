declare global {
  namespace webpack {
    interface Hot {
      monkeyReload(options?: import("./client/hmr").MonkeyReloadOptions): void
    }
  }
}

export * from "./node/monkey"
