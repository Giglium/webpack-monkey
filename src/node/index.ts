export * from "../shared/GM"
export * from "../shared/constants"
export * from "../shared/meta"
export * from "../shared/patching"
export * from "../shared/utils"

export * from "./monkey"
