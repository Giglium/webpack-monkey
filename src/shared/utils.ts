import { matchPattern } from "browser-extension-url-match"
import { isNil } from "lodash"

/**
 * array.includes() with type guard.
 */
export function includes<T>(array: readonly T[], value: any): value is T {
  return array.includes(value)
}

export function urlMatch(pattern: string, url: string) {
  const matcher = matchPattern(pattern)

  if (!matcher.valid) {
    throw new TypeError("Invalid pattern: " + pattern)
  }

  return matcher.match(url)
}

export function parentUntil<T>(
  target: T | null | undefined,
  getParent: (target: T) => T | null | undefined,
  callback: (target: T) => void | boolean
): T | null | undefined {
  const visited = new Set<T>()

  let current = target
  let result = null

  while (!isNil(current) && !visited.has(current)) {
    visited.add(current)

    if (callback(current)) {
      result = current
      return result
    }

    current = getParent(current)
  }

  return result
}
