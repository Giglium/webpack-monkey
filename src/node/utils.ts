// @ts-ignore
// prettier-ignore
import { getDescriptionFile, getRequiredVersionFromDescriptionFile } from "webpack/lib/sharing/utils";

import { isArray, isBoolean, isNil, isObject, isString, trimEnd } from "lodash"
import { Compilation, sources } from "webpack"
import { getGMAPIs } from "../shared/GM"
import { META_FIELDS, META_FIELDS_I18N, UserscriptMeta } from "../shared/meta"
import { includes } from "../shared/utils"

export function getPackageDepVersion(packageJson: any, dep: string): string | undefined {
  if (!isObject(packageJson)) {
    return undefined
  }

  return getRequiredVersionFromDescriptionFile(packageJson, dep)
}

export function getPackageJson(fs: Compilation["inputFileSystem"], context: string) {
  return new Promise<{ data: object; path: string }>((resolve, reject) => {
    getDescriptionFile(fs, context, ["package.json"], (err: any, result: any) => {
      if (err) {
        reject(err)
      } else {
        resolve(result)
      }
    })
  })
}

export function pathSplit(path: string) {
  return path.replace(/\\/g, "/").split("/")
}

export function traverseAndFindSource(
  source: unknown,
  cb: (source: sources.Source) => void | true,
): sources.Source | undefined {
  if (!source || !(source instanceof sources.Source)) {
    return undefined
  }

  const found = cb(source)

  if (found === true) {
    return source
  }

  if ((source as any)._children) {
    for (const child of (source as any)._children as unknown[]) {
      const result = traverseAndFindSource(child, cb)

      if (result) {
        return result
      }
    }
  }

  return traverseAndFindSource((source as any)._source, cb)
}

export function generateMetaBlock(source: string, meta: UserscriptMeta) {
  let metaBlock = "// ==UserScript==\n"
  const fieldPrefix = "// @"

  const metaFields = Object.keys(meta)
  const maxFieldLength = Math.max(
    ...["grant", "require", ...metaFields].map((field) => {
      if (includes(META_FIELDS_I18N, field)) {
        const value = meta[field]

        if (isObject(value)) {
          const maxLangLength = Math.max(
            ...Object.keys(value).map((lang) => (lang === "default" ? 0 : lang.length)),
          )

          return field.length + maxLangLength
        }
      }

      return field.length
    }),
  )
  const indentSize = fieldPrefix.length + maxFieldLength + 2
  const indentEnd = (str: string) => str.padEnd(indentSize, " ")

  function putField(field: string, value: string | string[] | boolean) {
    let line = fieldPrefix + field

    if (isString(value)) {
      line = indentEnd(line) + value
    } else if (isArray(value)) {
      if (!value.length) {
        return
      }

      line = value.map((v) => indentEnd(line) + v).join("\n")
    } else if (isBoolean(value)) {
      // ignore false value
      if (!value) {
        return
      }
    } else {
      console.warn("Unknown type of value:", value)
    }

    metaBlock += line + "\n"
  }

  for (const field of META_FIELDS) {
    if (includes(META_FIELDS_I18N, field)) {
      const value = meta[field]

      if (value) {
        if (isString(value)) {
          putField(field, value)
        } else {
          for (const lang of Object.keys(value)) {
            putField(field + (lang === "default" ? "" : ":" + lang), value[lang]!)
          }
        }
      }
    } else if (field === "grant") {
      putField(
        "grant",
        getGMAPIs().filter((api) => source.includes(api)),
      )

      if (meta.grant) {
        putField("grant", meta.grant)
      }
    } else {
      if (!isNil(meta[field])) {
        putField(field, meta[field]!)
      }
    }
  }

  metaBlock += "// ==/UserScript=="

  return metaBlock
}

export interface ResolveExternalResult {
  type?: string
  identifier?: string
  url: string
  value: string
}

export function resolveUrlExternal(externalValue: string): ResolveExternalResult | undefined {
  // full example: "var foo@https://example.com"
  const match = externalValue.match(/^(.*? )?(.+?@)?(https?:\/\/.+)$/)

  if (!match) {
    return undefined
  }

  const type = match[1]?.trim()
  const identifier = match[2] && trimEnd(match[2], "@")
  const url = match[3]!
  const value = (type ? type + " " : "") + (identifier || JSON.stringify(url))

  const resolved: ResolveExternalResult = {
    type,
    identifier,
    url,
    value,
  }

  return resolved
}

export function getUnnamedUrlExternalErrorMessage(url?: string) {
  const exampleUrl = url || "https://example.com"
  return `Unexpected reference to unnamed external module with URL "${url || "<unknown URL>"}".
This happens when you import a module from a URL
but do not specify an identifier for it,
e.g. \`import foo from "${exampleUrl}"\`, which will cause
runtime errors in the generated code. To fix this,
either add a universally unique identifier to the import statement,
e.g. \`import foo from "foo@${exampleUrl}"\`,
or do not import anything from it,
e.g. \`import "${exampleUrl}"\`.`.replace(/\n/g, " ")
}
