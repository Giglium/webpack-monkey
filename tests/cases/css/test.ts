import { it } from "@jest/globals"
import path from "path"
import { monkeyWebpack } from "../../../src"
import { testBuild, withCommonConfig } from "../../utils/webpack"

const config = withCommonConfig({
  entry: path.resolve(__dirname, "index.js"),
  output: {
    path: path.resolve(__dirname, "dist"),
  },
})

it("build", () => testBuild(monkeyWebpack()(config)))
