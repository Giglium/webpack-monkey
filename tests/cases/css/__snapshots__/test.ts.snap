// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build 1`] = `
"// ==UserScript==
// @name     CSS test
// @grant    GM_addStyle
// @match    *://*/*
// @version  1.0.0
// ==/UserScript==

;(() => {
  "use strict"

  // extracted by mini-css-extract-plugin
  /* harmony default export */ const styles_module = { tltle: "styles-module__tltle--zOaB" }

  console.log(styles_module.title)
})()

GM_addStyle(\`
body {
  background-color: red;
}

.styles-module__tltle--zOaB {
  color: white;
}

\`)
"
`;
